File: frontend/apps/staff/src/api/kitchenStream.ts
import { apiBaseUrl } from "./client";

/**
 * Agent.md §17: SSE from backend to KDS, "an optimization, not the only
 * source of truth" — events just tell the caller to refetch
 * `GET /api/v1/kitchen/orders`, they never carry state the UI trusts
 * directly. Reconnects automatically with a short backoff on any
 * disconnect, per §17's "browser reconnects automatically" requirement.
 *
 * Uses a manual fetch + ReadableStream reader rather than the browser's
 * `EventSource` API, because `EventSource` cannot send an `Authorization`
 * header and the kitchen stream is gated by the same staff bearer JWT as
 * every other `/api/v1/kitchen/**` endpoint (agent.md §14.1/§14.2).
 */
export function connectKitchenStream(token: string, onEvent: (eventName: string) => void): () => void {
  const controller = new AbortController();
  let stopped = false;

  async function run() {
    while (!stopped) {
      try {
        const response = await fetch(`${apiBaseUrl()}/api/v1/kitchen/stream`, {
          headers: { Authorization: `Bearer ${token}`, Accept: "text/event-stream" },
          signal: controller.signal,
        });
        if (!response.ok || !response.body) {
          throw new Error(`Kitchen stream responded with ${response.status}`);
        }

        const reader = response.body.getReader();
        const decoder = new TextDecoder();
        let buffer = "";

        while (!stopped) {
          const { value, done } = await reader.read();
          if (done) {
            break;
          }
          buffer += decoder.decode(value, { stream: true });

          let boundary = buffer.indexOf("\n\n");
          while (boundary !== -1) {
            const rawEvent = buffer.slice(0, boundary);
            buffer = buffer.slice(boundary + 2);
            const eventName = rawEvent
              .split("\n")
              .find((line) => line.startsWith("event:"))
              ?.slice("event:".length)
              .trim();
            if (eventName) {
              onEvent(eventName);
            }
            boundary = buffer.indexOf("\n\n");
          }
        }
      } catch {
        // Aborted on unmount, or the connection dropped — either way fall
        // through to the reconnect delay below (unless we were stopped).
      }

      if (!stopped) {
        await new Promise((resolve) => setTimeout(resolve, 3000));
      }
    }
  }

  void run();

  return () => {
    stopped = true;
    controller.abort();
  };
}

File: mobile/kitchen-android/app/src/main/java/com/example/pizzaconfigurator/kitchen/data/remote/KitchenStream.kt
package com.example.pizzaconfigurator.kitchen.data.remote

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import okhttp3.Request
import java.io.IOException

private const val RECONNECT_DELAY_MS = 3000L

/**
 * Ported 1:1 from the staff web app's `kitchenStream.ts`: `EventSource` cannot send an
 * `Authorization` header, and `/api/v1/kitchen/stream` is gated by the same staff bearer JWT as
 * every other staff-gated kitchen endpoint, so a hand-rolled streaming client is required here
 * too (OkHttp's `BufferedSource` standing in for the web version's `ReadableStream` reader).
 *
 * Only the `event:` line of each SSE frame is read — the `data:` field is never parsed. Per
 * agent.md §17, SSE is "an optimization, not the only source of truth": every event, regardless
 * of name, just means "something changed, refetch `GET /api/v1/kitchen/orders`." On any
 * disconnect (stream end or I/O error) this loops with a fixed 3s delay and reconnects, exactly
 * matching the web client's simple fixed backoff.
 */
class KitchenStream(private val bearerTokenProvider: suspend () -> String?) {

    suspend fun listen(onEvent: () -> Unit) {
        while (true) {
            try {
                connectOnce(onEvent)
            } catch (e: CancellationException) {
                throw e
            } catch (e: IOException) {
                // fall through to reconnect delay
            }
            delay(RECONNECT_DELAY_MS)
        }
    }

    private suspend fun connectOnce(onEvent: () -> Unit) = withContext(Dispatchers.IO) {
        val token = bearerTokenProvider() ?: return@withContext
        val request = Request.Builder()
            .url("${com.example.pizzaconfigurator.kitchen.BuildConfig.API_BASE_URL}/api/v1/kitchen/stream")
            .header("Authorization", "Bearer $token")
            .header("Accept", "text/event-stream")
            .build()

        NetworkFactory.okHttpClient.newCall(request).execute().use { response ->
            if (!response.isSuccessful) return@use
            val source = response.body?.source() ?: return@use
            val frame = StringBuilder()
            while (true) {
                val line = source.readUtf8Line() ?: break
                if (line.isEmpty()) {
                    val eventLine = frame.lineSequence().firstOrNull { it.startsWith("event:") }
                    if (eventLine != null) onEvent()
                    frame.clear()
                } else {
                    frame.appendLine(line)
                }
            }
        }
    }
}
